### PR TITLE
Waiting for review.

### LINKED ISSUE #3: Rework server
In the current state of server we're able to handle a single connection at the time. Instead we should operate on `set` of connected clients.

The task is to:
* Rework server to be able handle multiple connections at the same time
* Server should operate on all the connections (send a message for example)
* Server should be able to close connection with client that has no response

To be short, we should have server you can connect to, if client sends anything - respond with some message. In case client disconnected - remove connection from `set`.

### PR DESCRIPTION
**Pull request by unbreaking-V**

### CODE REVIEW

**Comment by ivysochyn** [comment]
Reword your commits to display changes they provide. Use the next naming scheme:
`[#<issue number>] <file/group was changed>: <file/group was changed>: Commit message`

For example:
`[#1] README.md: Fix typo`

In order to reword your commits use `git rebase`, inside editor mark corresponding commits with `r` instead of `pick`. In order to run rebase on your branch use:
`git rebase -i <commit hash>`

File: Makefile [hunk 1]
```diff
@@ -1,8 +1,14 @@
-main: server client
+TARGET = client server
+CC = g++  
+CFLAGS_OBJ =  -Wall -Wextra -pedantic -g 
+
+normal: $(TARGET)
```
Comment (ivysochyn) [inline Makefile:5]: `normal` should be renamed to `main`.

File: main.cpp [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+#include <vector>
+#include <string>
+
+#include "server.hpp"
+
+Server server;
+int main(int argc, char **argv)
+{
+    server.init();	
+	//actual main loop
```
Comment (ivysochyn) [inline main.cpp]: Replace all the tabs with four spaces, we should follow codestyle.

File: client.cpp [hunk 3]
```diff
@@ -1,21 +1,17 @@
-#include <sys/socket.h>
-#include <arpa/inet.h>
-#include <string.h>
-#include <unistd.h>
-#include <stdio.h>
+#include "server.hpp"
```
Comment (ivysochyn) [inline client.cpp:1]: Why are we using `server.hpp` to include? I believe we should have `client.cpp` independent of server.

Comment (ivysochyn) [reply client.cpp:1]: Too many spaces as well, follow the code style.

File: main.cpp [hunk 4]
```diff
@@ -0,0 +1,17 @@
+#include <iostream>
+#include <cstdint>
+#include <vector>
+#include <string>
+
```
Comment (ivysochyn) [inline main.cpp]: Is this new line is needed here?

File: server.cpp [hunk 5]
```diff
@@ -1,28 +1,189 @@
...
+
+
+void Server::setup(int port)
+{
+    if ((mastersocket = socket(AF_INET,SOCK_STREAM,IPPROTO_TCP)) <0){
+       perror("[SERVER] [ERROR] can't create a socket");
+       exit(EXIT_FAILURE);
+    }
+        
+    //Clear an masterfds and tempfds
```
Comment (ivysochyn) [inline server.cpp]: Use space after comment slashes: `// Text`

File: server.cpp [hunk 6]
```diff
@@ -1,28 +1,189 @@
...
-        socklen_t lClient = sizeof(client);
-        int connection = accept(server_socket, (sockaddr*)&client, &lClient);
-        char buf[128] = "Test if works\n";
-        write(connection, buf, strlen(buf));
-        shutdown(connection, SHUT_RDWR);
-        close(connection);
+    con_info.sin_addr.s_addr = htons(INADDR_ANY);
+    //zero the input buffer before use to avoid random data appearing in first receives
+    bzero(input_buffer,INPUT_BUFFER_SIZE);
+
```
Comment (ivysochyn) [inline server.cpp:44]: Inconsistent use of new line characters.

File: server.cpp [hunk 7]
```diff
@@ -1,28 +1,189 @@
...
+          
+}
+
+void Server::initializeSocket()
+{   
+    /*setsockopt - sets flags on the socket
```
Comment (ivysochyn) [inline server.cpp]: space

File: server.cpp [hunk 8]
```diff
@@ -1,28 +1,189 @@
...
+    //insert the master socket file-descriptor into the master fd-set
+	FD_SET(mastersocket, &masterfds);
+    //set the current known maximum file descriptor count;
+	maxfd = mastersocket;
+}
+
+void Server::startListen()
+{
+    /*Prepare a socket for peeling requests 
+     * and allows you to determine the size(3)
```
Comment (ivysochyn) [inline server.cpp]: Grammar errors

File: server.cpp [hunk 9]
```diff
@@ -1,28 +1,189 @@
...
+	int listen_ret = listen(mastersocket, 3);
+	if (listen_ret < 0) {
+		perror("[SERVER] [ERROR] listen() failed");
+	}
+
+}
+
+void Server::handleNewConnection()
+{   
+    socklen_t addrlen = sizeof (client_addr);
```
Comment (ivysochyn) [inline server.cpp]: Space between `sizeof` and `argument`.

File: server.cpp [hunk 10]
```diff
@@ -1,28 +1,189 @@
...
+            		maxfd = tempsocket;
+        	}
+    }
+}
+
+void Server::recvInputFromExisting(int fd)
+{   
+    int nbytesrecv = recv(fd, input_buffer, INPUT_BUFFER_SIZE, 0);
+    
+    printf("%s\n", input_buffer);
```
Comment (ivysochyn) [inline server.cpp]: Do we need to have this `printf` here?

File: server.cpp [hunk 11]
```diff
@@ -1,28 +1,189 @@
...
+    if (nbytesrecv <= 0)
+    {
+        /*If the client does not send any messages,
+         * it means that it has finished its work and left,
+         * and there is no need to maintain a connection with it
```
Comment (ivysochyn) [inline server.cpp]: Grammar

File: server.hpp [hunk 12]
```diff
@@ -0,0 +1,66 @@
+#ifndef SERVER_HPP
+#define SERVER_HPP 
+
+#include<cstdio>
+#include<cstdlib>
+#include<cstring>
+#include<fcntl.h>
```
Comment (ivysochyn) [inline server.hpp]: Do we really use all those headers?

File: server.hpp [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+        
+        //server socket details
+        struct sockaddr_in con_info;
+        //client connection data
+        struct sockaddr_storage client_addr;
+        
+        //input buffer 
+        char input_buffer[INPUT_BUFFER_SIZE];
+
+        void (*newConnectionCallback) (uint16_t fd);
```
Comment (ivysochyn) [inline server.hpp]: We're not using those callbacks, do we need them defined?

File: server.hpp [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+        void (*receiveCallback) (uint16_t fd, char *buffer);
+        void (*disconnectCallback) (uint16_t fd);
+
+
+        //function prototypes
+        void setup(int port);
+        void initializeSocket();
+        void bindSocket();
+        void startListen();
+        void shutdown();
```
Comment (ivysochyn) [inline server.hpp]: I don't implementation of shutdown, do we need it?

**Review by ivysochyn — CHANGES_REQUESTED** [review]
Follow code style with spaces/tabs, comments and brackets.
Review used headers to ensure we are not including not needed stuff.
Review not used/implemented functions.

File: main.cpp (hunk 4 above)
Comment (unbreaking-V) [reply main.cpp]: Remove include

File: server.cpp (hunk 10 above)
Comment (unbreaking-V) [reply server.cpp]: Removing printf

File: server.cpp (hunk 11 above)
Comment (unbreaking-V) [reply server.cpp]: Fixed

File: server.hpp (hunk 12 above)
Comment (unbreaking-V) [reply server.hpp]: Reviews include

File: server.hpp (hunk 13 above)
Comment (unbreaking-V) [reply server.hpp]: Removing callback function

File: server.hpp (hunk 14 above)
Comment (unbreaking-V) [reply server.hpp]: Remove shutdown function

**Review by ivysochyn — APPROVED** [review]

**Pull request merged**